internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,351 @@
+	// validate the volume ID
+	cr, err := util.NewUserCredentials(secrets)
+	if err != nil {
+		return nil, nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if volumeID == "" {
+		return nil, cr, status.Error(codes.InvalidArgument, "empty volume ID in request")
+	}
+
+	if acquired := cs.VolumeLocks.TryAcquire(volumeID); !acquired {
+		util.ErrorLog(ctx, util.VolumeOperationAlreadyExistsFmt, volumeID)
+		return nil, cr, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volumeID)
+	}
+
+	var rbdVol = &rbdVolume{}
+	rbdVol, err = genVolFromVolID(ctx, volumeID, cr, secrets)
+	if err != nil {
+		switch {
+		case errors.Is(err, ErrImageNotFound):
+			err = status.Errorf(codes.NotFound, "volume %s not found", volumeID)
+		case errors.Is(err, util.ErrPoolNotFound):
+			err = status.Errorf(codes.NotFound, "pool %s not found for %s", rbdVol.Pool, volumeID)
+		default:
+			err = status.Errorf(codes.Internal, err.Error())
+		}
+	}
+	return rbdVol, cr, err
+}
+
+// getForceOption extracts the force option from the GRPC request parameters.
+// If not set, the default will be set to false.
+func getForceOption(ctx context.Context, parameters map[string]string) (bool, error) {
+	val, ok := parameters[forceKey]
+	if !ok {
+		util.WarningLog(ctx, "%s is not set in parameters, setting to default (%v)", forceKey, false)
+		return false, nil
+	}
+	force, err := strconv.ParseBool(val)
+	if err != nil {
+		return false, status.Errorf(codes.Internal, err.Error())
+	}
+	return force, nil
+}
+
+// getMirroringMode gets the mirroring mode from the input GRPC request parameters.
+// mirroringMode is the key to check the mode in the parameters.
+func getMirroringMode(ctx context.Context, parameters map[string]string) (librbd.ImageMirrorMode, error) {
+	val, ok := parameters[imageMirroringKey]
+	if !ok {
+		util.WarningLog(ctx, "%s is not set in parameters, setting to mirroringMode to default (%s)", imageMirroringKey, imageMirrorModeSnapshot)
+		return librbd.ImageMirrorModeSnapshot, nil
+	}
+
+	var mirroringMode librbd.ImageMirrorMode
+	switch imageMirroringMode(val) {
+	case imageMirrorModeSnapshot:
+		mirroringMode = librbd.ImageMirrorModeSnapshot
+	case imageMirrorModeJournal:
+		mirroringMode = librbd.ImageMirrorModeJournal
+	default:
+		return mirroringMode, status.Errorf(codes.InvalidArgument, "%s %s not supported", imageMirroringKey, val)
+	}
+	return mirroringMode, nil
+}
+
+// cleanup performs below resource cleanup operations.
+func (cs *ControllerServer) cleanup(rbdVol *rbdVolume, cr *util.Credentials) {
+	if cr != nil {
+		// destroy the credential file
+		cr.DeleteCredentials()
+	}
+	if rbdVol != nil {
+		// destroy the cluster connection
+		rbdVol.Destroy()
+		// release the volume lock
+		cs.VolumeLocks.Release(rbdVol.VolID)
+	}
+}
+
+// EnableVolumeReplication enables replication on volume.
```

Comment:
A comment  for a public API function should describe what it does, and what conditions are important. The current comment is just phrasing the name of the function into human language. can you describe what errors are returned for what conditions?